man/tractorbeam.ronn
-*- mode: markdown; -*-

# tractorbeam(1) -- Monitors processes and reports state on zookeeper #

## SYNOPSIS ##

`tractorbeam` {send|recv} [OPTION]...

## DESCRIPTION ##

`tractorbeam` uses a user given program to collect information about
something and report its output on zookeeper.

The motivation for writing this was to use zookeeper as a naming
service. For instance, suppose you have a network service running on
several different machines. The the following would allow you to
discover these machines (suppose you run this on every single one):

    tractorbeam send \
                --zookeeper my-zk-cluster:2181 \
                --delay 1 \
                --path /my/service/$(uuid) \
                --exec /usr/bin/hostname \
                -- --fqdn

The above command would execute `hostname` command once a second
(`--delay` switch) then report the output on node given by the
`--path` switch.

Then you can read this entire tree and write it to a file:

    tractorbeam recv \
                --zookeeper my-zk-cluster:2181 \
                --path /my/service \
                --output zookeeper.tree

This would create/update an INI file with the contents of the
zookeeper cluster.

## RECV MODE ##

### SYNOPSIS ###

`tractorbeam` recv [OPTION]...

### OPTIONS ###

  * `--zookeeper` STRING:

    The zookeeper cluster you want to connect to. You may use comma to
    provide multiple hosts and optionally define a chroot path
    (e.g. "zk01:2181,zk02:2182/chroot");

  * `--path` STRING:

    The path of the tree you want to retrieve from zookeeper;

  * `--output` {PATH}:

    The file to write contents into (layout=file) or the directory to
    create the zk tree (layout=filesystem). The value `-` means
    stdout when using layout=file;

  * `--layout` {filesystem,file}:

    The layout to use when reading the zookeeper tree.

    The `filesystem` layout uses files and directories. For instance:

        zkCli $ create /foo foo
        zkCli $ create /foo/bar foobar
        $ tractorbeam --layout filesystem --path /foo --output /tmp/zk
        $ find /tmp/zk
        /tmp/zk
        /tmp/zk/foo
        /tmp/zk/foo.data      # <- contents of /foo
        /tmp/zk/foo/bar
        /tmp/zk/foo/bar.data  # <- contents of /foo/bar

    The `file` layout creates a single file, using the following format:

        zkCli $ create /foo foo
        zkCli $ create /foo/bar foobar
        $ tractorbeam --layout file --path /foo --output /tmp/zk.data
        $ cat /tmp/zk.data
        /foo|3
        foo
        /foo/bar|6
        foobar

     The format of this file is as follows:

        <PATH> "|" <SIZE> "\n"
        <CONTENTS> "\n"
       
## SEND MODE ##

### SYNOPSIS ###

`tractorbeam` send [OPTION]... -- [ARGV]...

### OPTIONS ###

  * `--zookeeper` STRING:

    The zookeeper cluster you want to connect to. You may use comma to
    provide multiple hosts and optionally define a chroot path
    (e.g. "zk01:2181,zk02:2182/chroot");

  * `--path` STRING:

    The path of the node you want to create on zookeeper. Notice all
    the parents must exist and you must have write permissions.

    The node is *ephemeral* and have *OPEN_UNSAFE* acl;

  * `--exec` FILE:

    The image you want to execute. This must be an absolute path. This
    program must terminate faster than `--delay` otherwise it
    gets killed and the node on zookeeper is removed.

    The data you want to write on zookeeper should be written on
    *stdout* (a pipe). Additionally *stdin* and *stderr* are
    redirected to */dev/null*.

    If this process *exists 0* the zookeeper node gets either created
    or updated. Anything else deletes the node;

  * `--delay` SECONDS:

    The interval at which the `--exec` program gets invoked (the time
    spent running the process is not taken into account). This value
    also defines the timeout for the process to terminate;

  * `--timeout` MILLISECS:

    The timeout option to use when connecting zookeeper. This informs
    zookeeper to keep the ephemeral this long before deleting in the
    event no heartbeat is received;

  * `--help`:

    Prints a short help message;

## AUTHOR ##

Written by dgvncsz0f

## COPYRIGHT ##

Copyright (c) 2013 Diego Souza. License BSD.

## TODO ##

  * Sign the contents of the mode;


CHANGELOG
-*- mode: text; -*-

v1.1.0 | 2013-11-06
  * recv filesystem layout | dgvncsz0f

v1.0.0 | 2013-11-04
  * creating recv mode (reading zookeeper tree) | dgvncsz0f
  * major: cli interface has changed: adding subcommands | dgvncsz0f

v0.1.0 | 2013-11-01
  * first version | dgvncsz0f


src/tractorbeam.c
// All rights reserved.
//  
// Redistribution and use in source and binary forms, with or without modification,
// are permitted provided that the following conditions are met:
//  
// * Redistributions of source code must retain the above copyright notice, this
//   list of conditions and the following disclaimer.
//  
// * Redistributions in binary form must reproduce the above copyright notice, this
//   list of conditions and the following disclaimer in the documentation and/or
//   other materials provided with the distribution.
//  
// * Neither the name of the {organization} nor the names of its
//   contributors may be used to endorse or promote products derived from
//   this software without specific prior written permission.
//  
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
// ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#include <stdio.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <string.h>
#include "tractorbeam/debug.h"
#include "tractorbeam/zkrecv.h"
#include "tractorbeam/zksend.h"
#include "tractorbeam/helpers.h"
#include "tractorbeam/monitor.h"

#define TB_DEFAULT_ENDPOINT "localhost:2181"
#define TB_DEFAULT_TIMEOUT 5000
#define TB_DEFAULT_DELAY 5
#define TB_RECV_BUFSIZE 2097152

static
int __tractorbeam_check_send(tractorbeam_zksend_t *sendcfg)
{
  int rc = 0;
  if (sendcfg->endpoint == NULL || strcmp("", sendcfg->endpoint) == 0)
  {
    printf("ERROR: zookeeper must not be null\n");
    rc = 1;
  }

  if (sendcfg->path == NULL || strcmp("", sendcfg->path) == 0)
  {
    printf("ERROR: path must not be null\n");
    rc = 1;
  }

  if (sendcfg->exec == NULL || strcmp("", sendcfg->exec) == 0)
  {
    printf("ERROR: exec must not be null\n");
    rc = 1;
  }

  if (sendcfg->timeout <= 0)
  {
    printf("ERROR: timeout must be >0\n");
    rc = 1;
  }

  return(rc);
}

static
int __tractorbeam_check_recv(tractorbeam_zkrecv_t *recvcfg)
{
  int rc = 0;

  if (recvcfg->path == NULL || strcmp("", recvcfg->path) == 0)
  {
    printf("ERROR: path must not be null\n");
    rc = 1;
  }

  return(rc);
}

static
size_t __tractorbeam_strlen1(const char *s)
{
  size_t maxlen = 0;
  size_t curlen = 0;
  for (; s[0] != '\0'; s++)
  {
    curlen += 1;
    if (s[0] == '\n')
    {
      maxlen = (curlen > maxlen) ? curlen : maxlen;
      curlen = 0;
    }
  }

  return(curlen > maxlen ? curlen : maxlen);
}

static
void __printf_indent(const char *a, const char *b, size_t maxlen)
{
  char *tmp     = tbh_strdup(b);
  char *delim   = " ";
  char *token   = NULL;
  size_t prefix = __tractorbeam_strlen1(a);
  size_t at     = 0;

  printf("\n%s", a);
  token = strtok(tmp, delim);
  do
  {
    at += strlen(token);
    if ((at+prefix > maxlen) || (strcmp("\n", token) == 0))
    {
      at = prefix;
      printf("\n");
      while (at-- > 0)
      { printf(delim); }
      at = 0;
    }
    if (strcmp("\n", token) != 0)
    { printf("%s ", token); }
  } while ((token = strtok(NULL, delim)) != NULL);
  printf("\n");

  free(tmp);
}

static
void __tractorbeam_print_usage0(const char *prg)
{
  printf("USAGE: %s {send,recv} OPTIONS...\n\n", prg);
  printf("  tip: use --help after the sub-comamnd to get a list of available options\n");
}

static
void __tractorbeam_print_sendusage(const char *prg)
{
  char buffer[1024];
  printf("USAGE: %s send OPTIONS... -- [ARGV...]\n", prg);

  __printf_indent("", "  This program connects to a zookeeper cluster and regularly writes"
                      "  the output of a given program into a ephemeral node.", 60);

  snprintf(buffer, 1024, "The zookeeper cluster to connect to [default:%s];", TB_DEFAULT_ENDPOINT);
  __printf_indent("  --zookeeper STRING  ", buffer, 76);

  snprintf(buffer, 1024, "The path of the ephemeral node you want to create;");
  __printf_indent("  --path STRING       ", buffer, 76);

  snprintf(buffer, 1024, "The image to invoke. This should be the an absolute path (but it is"
                         " not enforced);");
  __printf_indent("  --exec FILE         ", buffer, 76);

  snprintf(buffer, 1024, "Defines the interval at which the program gets called. The amount of"
                         " time the program spent during its thing is not taken into account,"
                         " but the runtime may not exceed this value [default:%d];", TB_DEFAULT_DELAY);
  __printf_indent("  --delay SECONDS     ", buffer, 76);

  snprintf(buffer, 1024, "This defines how much time without communication zookeeper should"
                         " consider the client still alive [default:%d];", TB_DEFAULT_TIMEOUT);
  __printf_indent("  --timeout MILLISECS ", buffer, 76);

}

static
void __tractorbeam_print_recvusage(const char *prg)
{
  char buffer[1024];
  printf("USAGE: %s recv OPTIONS...\n", prg);

  __printf_indent("", "  This program connects to a zookeeper cluster and reads its"
                      "  entire tree allowing the user to dump the contents in many"
                      "  different formats.", 60);

  snprintf(buffer, 1024, "The zookeeper cluster to connect to [default:%s];", TB_DEFAULT_ENDPOINT);
  __printf_indent("  --zookeeper STRING         ", buffer, 76);

  snprintf(buffer, 1024, "The tree you want to read;");
  __printf_indent("  --path STRING              ", buffer, 76);

  snprintf(buffer, 1024, "The file to write the contents. Use - to write into the stdout [default:-];");
  __printf_indent("  --output FILE              ", buffer, 76);

  snprintf(buffer, 1024, "The layout to use when dumping the zookeeper tree. `filesystem' uses"
                         " files and directories and `file' uses a single file [default:file];\n");
  __printf_indent("  --layout {filesystem,file} ", buffer, 76);
}

static
int __tractorbeam_parse_recvopts(int argc, char *argv[], tractorbeam_zkrecv_t *recvcfg)
{
  static struct option my_options[] = {
    {"zookeeper",     required_argument, NULL, 0 },
    {"path",          required_argument, NULL, 0 },
    {"output",        required_argument, NULL, 0 },
    {"layout",        required_argument, NULL, 0 },
    {"help",          no_argument,       NULL, 0 },
    {0,               0,                 NULL, 0 }
  };

  while (1)
  {
    int opt = 0;
    int rc  = getopt_long_only(argc, argv, "", my_options, &opt);
    if (rc == -1)
    { break; }
    else if (rc == 0)
    {
      if (opt == 0)
      { recvcfg->endpoint = optarg; }
      else if (opt == 1)
      { recvcfg->path = optarg; }
      else if (opt == 2)
      { recvcfg->output = optarg; }
      else if (opt == 3)
      {
        if (strcmp("file", optarg) == 0)
        { recvcfg->layout = ZKRECV_LAYOUT_FILE; }
        else if (strcmp("filesystem", optarg) == 0)
        { recvcfg->layout = ZKRECV_LAYOUT_FILESYSTEM; }
        else
        {
          printf("ERROR: invalid layout\n");
          return(-1);
        }
      }
      else
      { return(-1); }
    }
    else
    { return(-1); }
  }

  return(__tractorbeam_check_recv(recvcfg));
}

static
int __tractorbeam_parse_sendopts(int argc, char *argv[], tractorbeam_zksend_t *sendcfg)
{
  static struct option my_options[] = {
    {"zookeeper",     required_argument, NULL, 0 },
    {"path",          required_argument, NULL, 0 },
    {"exec",          required_argument, NULL, 0 },
    {"timeout",       required_argument, NULL, 0 },
    {"delay",         required_argument, NULL, 0 },
    {"help",          no_argument,       NULL, 0 },
    {0,               0,                 NULL, 0 }
  };

  while (1)
  {
    int opt = 0;
    int rc  = getopt_long_only(argc, argv, "", my_options, &opt);
    if (rc == -1)
    { break; }
    else if (rc == 0)
    {
      if (opt == 0)
      { sendcfg->endpoint = optarg; }
      else if (opt == 1)
      { sendcfg->path = optarg; }
      else if (opt == 2)
      { sendcfg->exec = optarg; }
      else if (opt == 3)
      { sendcfg->timeout = atoi(optarg); }
      else if (opt == 4)
      { sendcfg->delay = atoi(optarg); }
      else
      { return(-1); }
    }
    else
    { return(-1); }
  }

  int k            = optind;
  sendcfg->argv    = (char **) malloc(sizeof(char*) * (2 + (argc - optind)));
  sendcfg->argv[0] = sendcfg->exec;
  for (; k<argc; k+=1)
  { sendcfg->argv[1+k-optind] = argv[k]; }
  sendcfg->argv[1+k-optind] = NULL;

  return(__tractorbeam_check_send(sendcfg));
}

int main(int argc, char *argv[])
{
  tractorbeam_zksend_t sendcfg;
  sendcfg.endpoint  = TB_DEFAULT_ENDPOINT;
  sendcfg.path      = "";
  sendcfg.exec      = "";
  sendcfg.argv      = NULL;
  sendcfg.delay     = TB_DEFAULT_DELAY;
  sendcfg.timeout   = TB_DEFAULT_TIMEOUT;

  tractorbeam_zkrecv_t recvcfg;
  recvcfg.endpoint  = TB_DEFAULT_ENDPOINT;
  recvcfg.path      = "";
  recvcfg.output    = "";
  recvcfg.layout    = ZKRECV_LAYOUT_FILE;
  recvcfg.delay     = TB_DEFAULT_DELAY;
  recvcfg.timeout   = TB_DEFAULT_TIMEOUT;

  if (argc < 2)
  {
    __tractorbeam_print_usage0(argv[0]);
    return(-1);
  }

  if (strcmp("send", argv[1]) == 0)
  {
    argv[1] = argv[0];
    if (__tractorbeam_parse_sendopts(argc-1, argv+1, &sendcfg) != 0)
    {
      __tractorbeam_print_sendusage(argv[0]);
      return(-1);
    }

    int rc = tractorbeam_zksend(&sendcfg);
    free(sendcfg.argv);
    return(rc);
  }
  else if (strcmp("recv", argv[1]) == 0)
  {
    argv[1] = argv[0];
    if (__tractorbeam_parse_recvopts(argc-1, argv+1, &recvcfg) != 0)
    {
      __tractorbeam_print_recvusage(argv[0]);
      return(-1);
    }

    return(tractorbeam_zkrecv(&recvcfg));
  }
  else
  {
    __tractorbeam_print_usage0(argv[0]);
    return(-1);
  }

  return(-1);
}
